calico_node/allocateipip/allocate_ipip_addr.go

```diff
@@ -74,7 +74,7 @@ func ensureHostTunnelAddress(ctx context.Context, c client.Interface, nodename s
 		log.WithError(err).Fatalf("Unable to retrieve IPIP tunnel address. Error getting node '%s'", nodename)
 	}
 
-	if node.Spec.BGP.IPv4IPIPTunnelAddr == "" {
+	if node.Spec.BGP == nil || node.Spec.BGP.IPv4IPIPTunnelAddr == "" {
```

Comment:
Hm, I wonder if we ever want to assign an IPIP tunnel address if the BGP spec is `nil` - seems like in this case we really want to remove any tunnel address (since BGP isn't being used, and thus IPIP won't be either).

The cases where BGP is nil:
- Using canal or something other than Calico BGP for routing.
- When IPIP is disabled and a BGP IP address conflict is detected, BGP spec will be temporarily nil

WDYT?

Reply:
I'm not sure I follow what you are suggesting.

> if the BGP spec is nil - seems like in this case we really want to remove any tunnel address

If BGP spec is nil then we don't have the tunnel address to remove.

> When IPIP is disabled and a BGP IP address conflict is detected, BGP spec may be temporarily nil

I don't think this code will be executed because ensureHostTunnelAddress is only called when there is an IPIP enabled pool.

Reply:
> I don't think this code will be executed because ensureHostTunnelAddress is only called when there is an IPIP enabled pool.

Er, yeah I meant when `node.Spec.BGP.IPv4IPIPTunnelAddr == ""`, sometimes the BGP spec will be set to nil. I think we can ignore that scenario though because we won't reach this code.

> If BGP spec is nil then we don't have the tunnel address to remove.

Well, technically it might still be allocated in IPAM and we'd want to remove it, but I don't think we've got the infra in place to do that since right now we require having the IP in-hand in order to free it.

I guess my point was that if the BGP spec is `nil`, then why would we assign a tunnel address? I don't _think_ there are any cases where the BGP spec is nil and we'd want to assign a tunnel address.

That said, I don't think it really hurts, so I'm happy to leave it as it's written right now and see if we get feedback.

Reply:
Like we talked about I'll add a new `if` branch for if BGP spec is nil to log and then return back.